components/timezone/src/types.rs

```diff
@@ -35,6 +40,79 @@ impl GmtOffset {
         Self(0)
     }
 
+    /// Parse a [`GmtOffset`] from bytes.
+    ///
+    /// The offset must range from GMT-12 to GMT+14.
+    /// The string must be an ISO-8601 time zone designator:
+    /// e.g. Z
+    /// e.g. +05
+    /// e.g. +0500
+    /// e.g. +05:00
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::timezone::GmtOffset;
+    /// use icu::timezone::TimeZoneError;
+    ///
+    /// let offset0: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"Z").expect("Failed to parse a time zone");
+    /// let offset1: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+05").expect("Failed to parse a time zone");
+    /// let offset2: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+0500").expect("Failed to parse a time zone");
+    /// let offset3: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"-05:00").expect("Failed to parse a time zone");
+    /// let offset_err0: TimeZoneError =
+    ///     GmtOffset::try_from_bytes(b"0500").expect_err("Invalid input");
+    /// let offset_err1: TimeZoneError =
+    ///     GmtOffset::try_from_bytes(b"+05000").expect_err("Invalid input");
+    ///
+    /// assert_eq!(offset0.offset_seconds(), 0);
+    /// assert_eq!(offset1.offset_seconds(), 18000);
+    /// assert_eq!(offset2.offset_seconds(), 18000);
+    /// assert_eq!(offset3.offset_seconds(), -18000);
+    /// assert_eq!(offset_err0, TimeZoneError::InvalidOffset);
+    /// assert_eq!(offset_err1, TimeZoneError::InvalidOffset);
+    /// ```
+    pub fn try_from_bytes(chars: &[u8]) -> Result<Self, TimeZoneError> {
+        let mut it = chars.iter();
+        let offset_sign = match it.next() {
+            Some(b'+') => 1,
+            Some(b'-') => -1,
```

Comment:
Now that we have bytes, we don't need to go through an iterator anymore. You can pattern match here as well, which lets you preserve the `\u{2212}` compatibility:
```rust
    let offset_sign = match chars {
        [b'+', rest @ ..] => {
            chars = rest;
            1
        }
        [b'-', rest @ ..] => {
            chars = rest;
            -1
        }
        // Unicode minus
        [226, 136, 146, rest @ ..] => {
            chars = rest;
            -1
        }
        [b'Z'] => return Ok(Self(0)), // this is actually a bug fix, before it accepted "Zanything"
        _ => return Err(TimeZoneError::InvalidOffset),
    };
```

Reply:
True, it's more explicit than an iterator. I will play with the compiler to learn more about the cost and slice pattern matching. I will add your comment about `Z` case.

@Manishearth does the API intend to support Unicode minus? I feel that users should convert/construct their input to ASCII `-` with timezones. At least, I haven't seen people attempting a timestamp-styled timezone with Unicode minus...

Reply:
Uh, I don't know, @nordzilla would